metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py

```diff
@@ -0,0 +1,138 @@
+import logging
+from typing import Dict, List, Optional
+
+from datahub.ingestion.source.snowflake.snowflake_config import (
+    SnowflakeV2Config,
+    TagOption,
+)
+from datahub.ingestion.source.snowflake.snowflake_report import SnowflakeV2Report
+from datahub.ingestion.source.snowflake.snowflake_schema import (
+    SnowflakeDataDictionary,
+    SnowflakeTag,
+    _SnowflakeTagCache,
+)
+from datahub.ingestion.source.snowflake.snowflake_utils import SnowflakeCommonMixin
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class SnowflakeTagExtractor(SnowflakeCommonMixin):
+    def __init__(
+        self,
+        config: SnowflakeV2Config,
+        data_dictionary: SnowflakeDataDictionary,
+        report: SnowflakeV2Report,
+    ) -> None:
+        self.config = config
+        self.data_dictionary = data_dictionary
+        self.report = report
+        self.logger = logger
+
+        self.tag_cache: Dict[str, _SnowflakeTagCache] = {}
+
+    def get_tags_on_object(
+        self,
+        domain: str,
+        db_name: str,
+        schema_name: Optional[str] = None,
+        table_name: Optional[str] = None,
+    ) -> List[SnowflakeTag]:
+        if self.config.extract_tags == TagOption.without_lineage:
+            if db_name not in self.tag_cache:
+                self.tag_cache[
+                    db_name
+                ] = self.data_dictionary.get_tags_for_database_without_propagation(
+                    db_name
+                )
+
+            if domain == "database":
+                return self.tag_cache[db_name].get_database_tags(db_name)
+            elif domain == "schema":
+                assert schema_name is not None
+                tags = self.tag_cache[db_name].get_schema_tags(schema_name, db_name)
+            elif domain == "table":  # Views belong to this domain as well.
+                assert schema_name is not None
+                assert table_name is not None
+                tags = self.tag_cache[db_name].get_table_tags(
+                    table_name, schema_name, db_name
+                )
+            else:
+                raise ValueError(f"Unknown domain {domain}")
+        elif self.config.extract_tags == TagOption.with_lineage:
+            identifier = ""
```

Comment:
Can we  refractor contents of else condition into another method.  Say -  `get_tags_on_object_with_lineage`